[BUG] Buffer prefetching off by one, prefetch=1 doesn't prefetch at all, since v0.3

## Describe the bug

In torchrl.data.ReplayBuffer, the multithreaded prefetching logic has an off-by-one error. When prefetch=N is set, the internal prefetch queue (_prefetch_queue) only ever maintains N-1 items in the background after yielding a batch. For example, if a user sets prefetch=1, the queue fills to 1, is immediately popped, and 0 items are left prefetching in the background.

## To Reproduce

```python
import torch
from torchrl.data import ReplayBuffer, ListStorage

def test_prefetch_bug():
    # Initialize a replay buffer with prefetch=1
    rb = ReplayBuffer(
        storage=ListStorage(max_size=100),
        batch_size=2,
        prefetch=1
    )

    # Populate it with dummy tensor data
    rb.extend(torch.arange(100))

    # Trigger the first sample.
    # The user expects 1 task to be left fetching in the background.
    _ = rb.sample()

    # Inspect the internal prefetch queue length.
    queue_length = len(rb._prefetch_queue)
    
    print(f"Prefetch queue length: {queue_length}")
    
    # Under current behavior, this assert fails because queue_length == 0
    assert queue_length == 1, f"Bug present: Expected prefetch queue to have 1 item, but got {queue_length}."

if __name__ == "__main__":
    test_prefetch_bug()
```

```bash
Traceback (most recent call last):
  File "mre.py", line 28, in <module>
    test_prefetch_bug()
  File "mre.py", line 25, in test_prefetch_bug
    assert queue_length == 1, f"Bug present: Expected prefetch queue to have 1 item, but got {queue_length}."
AssertionError: Bug present: Expected prefetch queue to have 1 item, but got 0.
```

## Expected behavior

Setting prefetch=N should maintain exactly N batches computing in the background after the sample() method returns.

## System info

0.13.2 2.4.6 3.14.6 | packaged by conda-forge | (main, Jun 12 2026, 08:51:42) [GCC 14.3.0] linux

## Reason and Possible fixes

Currently, popleft().result() is called immediately after the while loop that fills the queue. This inevitably leaves the queue with _prefetch_cap - 1 items.

We can resolve both the off-by-one error and the initial dispatch overhead by evaluating the immediate need synchronously when the queue is empty, and then topping up the prefetch queue to exactly _prefetch_cap.

In torchrl/data/replay_buffers/replay_buffers.py under the sample method:

Current code:
```python
else:
            with self._futures_lock:
                while (
                    len(self._prefetch_queue)
                    < min(self._sampler._remaining_batches, self._prefetch_cap)
                    and not self._sampler.ran_out
                ) or not len(self._prefetch_queue):
                    fut = self._prefetch_executor.submit(self._sample, batch_size)
                    self._prefetch_queue.append(fut)
                result = self._prefetch_queue.popleft().result()
```

Proposed fix:
```python
else:
            with self._futures_lock:
                if len(self._prefetch_queue):
                    result = self._prefetch_queue.popleft().result()
                else:
                    result = self._sample(batch_size)
                while (
                    len(self._prefetch_queue)
                    < min(self._sampler._remaining_batches, self._prefetch_cap)
                    and not self._sampler.ran_out
                ):
                    fut = self._prefetch_executor.submit(self._sample, batch_size)
                    self._prefetch_queue.append(fut)
```
I plan to submit a PR for this fix shortly.
The issue seems to have been created in [this commit](https://github.com/pytorch/rl/commit/4d3a0c6c1fb99c58023de4da03be4901419a4457) in v0.3.
## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

--- a/test/rb/test_rb_core.py
+++ b/test/rb/test_rb_core.py
@@ -623,6 +623,30 @@ def test_replay_buffer_iter(size, drop_last):
         assert i == (size - 1) // 3
 
 
+def test_replay_buffer_prefetch_queue_length():
+    """Test that the prefetch queue maintains the correct length.
+    
+    This test verifies that after sampling from a replay buffer with prefetching
+    enabled, the prefetch queue has exactly `prefetch` items computing in the
+    background (no off-by-one error).
+    """
+    torch.manual_seed(0)
+    
+    rb = ReplayBuffer(
+        storage=ListStorage(max_size=100),
+        batch_size=2,
+        prefetch=2
+    )
+
+    rb.extend(torch.arange(100))
+
+    _ = rb.sample()
+
+    assert len(rb._prefetch_queue) == 2, (
+        f"Expected prefetch queue to have 2 items, but got {len(rb._prefetch_queue)}."
+    )
+
+
 if __name__ == "__main__":
     args, unknown = argparse.ArgumentParser().parse_known_args()
     pytest.main([__file__, "--capture", "no", "--exitfirst"] + unknown)

--- a/test/rb/test_rb_core.py
+++ b/test/rb/test_rb_core.py
@@ -625,26 +625,22 @@ def test_replay_buffer_iter(size, drop_last):
 
 def test_replay_buffer_prefetch_queue_length():
     """Test that the prefetch queue maintains the correct length.
-    
+
     This test verifies that after sampling from a replay buffer with prefetching
     enabled, the prefetch queue has exactly `prefetch` items computing in the
     background (no off-by-one error).
     """
     torch.manual_seed(0)
-    
-    rb = ReplayBuffer(
-        storage=ListStorage(max_size=100),
-        batch_size=2,
-        prefetch=2
-    )
+
+    rb = ReplayBuffer(storage=ListStorage(max_size=100), batch_size=2, prefetch=2)
 
     rb.extend(torch.arange(100))
 
     _ = rb.sample()
 
-    assert len(rb._prefetch_queue) == 2, (
-        f"Expected prefetch queue to have 2 items, but got {len(rb._prefetch_queue)}."
-    )
+    assert (
+        len(rb._prefetch_queue) == 2
+    ), f"Expected prefetch queue to have 2 items, but got {len(rb._prefetch_queue)}."
 
 
 if __name__ == "__main__":

--- a/torchrl/data/replay_buffers/replay_buffers.py
+++ b/torchrl/data/replay_buffers/replay_buffers.py
@@ -1154,14 +1154,17 @@ class ReplayBuffer:
             result = self._sample(batch_size)
         else:
             with self._futures_lock:
+                if len(self._prefetch_queue):
+                    result = self._prefetch_queue.popleft().result()
+                else:
+                    result = self._sample(batch_size)
                 while (
                     len(self._prefetch_queue)
                     < min(self._sampler._remaining_batches, self._prefetch_cap)
                     and not self._sampler.ran_out
-                ) or not len(self._prefetch_queue):
+                ):
                     fut = self._prefetch_executor.submit(self._sample, batch_size)
                     self._prefetch_queue.append(fut)
-                result = self._prefetch_queue.popleft().result()
 
         if return_info:
             out, info = result

